newIDE/app/src/UI/Menu/ElementWithMenu.js

```diff
@@ -5,45 +5,78 @@ import ContextMenu from './ContextMenu';
 
 type Props = {|
   element: React$Element<any>,
-  buildMenuTemplate: () => Array<any>,
+  onClick?: () => void,
+  buildMenuTemplate?: () => Array<any>,
+  buildMenuTemplateRight?: () => Array<any>,
 |};
 
 type State = {||};
 
 /**
  * Wrap an element and display a menu when `onClick` prop is called on the element.
  */
+
 export default class ElementWithMenu extends React.Component<Props, State> {
   _contextMenu: ?ContextMenu;
+  _contextMenuR: ?ContextMenu;
   _wrappedElement: ?any;
 
-  open = (event: any) => {
-    const { _contextMenu } = this;
-    if (!_contextMenu) return;
+  open = (element: any) => {
+    if (!this.props.buildMenuTemplate) {
+      this.props.element.props.onClick();
+    } else {
+      const { _contextMenu } = this;
+      if (!_contextMenu) return;
+
+      const node = ReactDOM.findDOMNode(this._wrappedElement);
+      if (node instanceof HTMLElement) {
+        const dimensions = node.getBoundingClientRect();
+
+        _contextMenu.open(
+          Math.round(dimensions.left + dimensions.width / 2),
+          Math.round(dimensions.top + dimensions.height)
+        );
+      }
+    }
+  };
+
+  openRight = () => {
+    const { _contextMenuR } = this;
+    if (!_contextMenuR) return;
 
     const node = ReactDOM.findDOMNode(this._wrappedElement);
     if (node instanceof HTMLElement) {
       const dimensions = node.getBoundingClientRect();
 
-      _contextMenu.open(
-        Math.round(dimensions.left),
+      _contextMenuR.open(
+        Math.round(dimensions.left + dimensions.width / 2),
         Math.round(dimensions.top + dimensions.height)
       );
     }
   };
 
   render() {
-    const { element, buildMenuTemplate } = this.props;
+    const { element, buildMenuTemplate, buildMenuTemplateRight } = this.props;
+
     return (
       <React.Fragment>
         {React.cloneElement(element, {
           onClick: this.open,
+          onContextMenu: this.openRight,
```

Comment:
Here instead of `onClick: this.open`, you can do:

`const { openMenuWithSecondaryClick, ... } = this.props`

```
React.cloneElement(element, {
  ...(openMenuWithSecondaryClick ? {onContextMenu: this.open} : {onClick: this.open}),
  ref: wrappedElement => (this._wrappedElement = wrappedElement),
}
```

(the weird `...()` is to make sure that we don't override the existing `onClick` in case we have an element that is defining onClick already)